File: javaNeuralNetwork/src/main/java/neural_network/util/WeightedPartitioner.java
package neural_network.util;

import java.util.*;
import java.util.stream.Stream;

/** Class to create a number of sets of size {@code setSize} from a list of
 * {@code numInts} integers weighted by which ground truth class each integer lies in.
 *
 */
public class WeightedPartitioner extends Partitioner {

    private int numBins;
    private final Map<Integer, List<Integer>> classMap = new HashMap<>();

    /** Constructor method.
     *
     * @param numInts Number of integers.
     * @param setSize Size of each set.
     * @param yVals The classes/values for each datapoint.
     * @param doRegression Whether we have a regression or classification problem.
     * @param numBins If {@code doRegression} is {@code true}, this represents the
     *                number of bins to split the data into. Otherwise, this
     *                parameter is ignored.
     */
    public WeightedPartitioner(int numInts, int setSize, List<Double> yVals,
                               boolean doRegression, int numBins) {
        super(numInts, setSize);
        if (! (yVals.size() == numInts)) {
            throw new IllegalArgumentException("numInts must equal the length of the " +
                    "yVals (numInts = %d, yVals.size() = %d)"
                            .formatted(numInts, yVals.size()));
        }

        // For regression, we split data into bins
        if (doRegression) {
            this.numBins = numBins;
            NavigableMap<Integer, List<Integer>> initialMap = new TreeMap<>();
            Stream.iterate(0, j -> j < numBins, j -> j + 1)
                    .forEach(j -> initialMap.put(j, new ArrayList<>(List.of())));
            double minY = Collections.min(yVals); double maxY = Collections.max(yVals);
            double classWidth = (maxY - minY) / numBins;
            for (int i = 0; i < yVals.size(); i ++) {

                // Here, we add the index of the y value to a specific discrete bin,
                // depending on where it lies in the range (minY, maxY)
                double y = yVals.get(i);
                int chosenBin = (int) ((y - minY) / classWidth);
                if (chosenBin == numBins) {
                    chosenBin = numBins - 1;
                }
                initialMap.get(chosenBin).add(i);
            }

            // We also need to account for potentially empty class lists
            for (int j = 0; j < this.numBins; j ++) {
                if (initialMap.get(j).isEmpty()) {
                    // Remove the bin and reduce numBins
                    initialMap.remove(j);
                    this.numBins --;
                }
            }

            // Finally, relabel the bins to 0, 1, 2, ...
            for (int k = 0; k < this.numBins; k ++) {
                // The below line polls the first key in the initialMap and uses it
                // to retrieve the list of integers at this entry. Then, this is put
                // in the new classMap with relabelled key
                classMap.put(k, initialMap.get(initialMap.firstKey()));
                initialMap.remove(initialMap.firstKey());
            }
        } else {
            // For classification, this process just involves getting the classes
            List<Integer> yClasses = yVals.stream()
                    .map(x -> (int) (double) x)
                    .toList();
            this.numBins = (new HashSet<>(yVals)).size();
            for (int i = 0; i < yClasses.size(); i ++) {
                classMap.computeIfAbsent(yClasses.get(i),
                        k -> new ArrayList<>()).add(i);
            }
        }
    }

    /** Constructor for classification. Note that {@code numBins} will not be used
     * in a classification problem.
     *
     * @param numInts Number of integers.
     * @param setSize Size of each set.
     * @param yVals The classes/values for each datapoint.
     */
    public WeightedPartitioner(int numInts, int setSize, List<Double> yVals) {
        this(numInts, setSize, yVals, false, 10);
    }

    /** Uses weights for each class to create sets of size {@code setSize} containing
     * integers (sampled with replacement).
     *
     * @return The list of lists of datapoint indices.
     */
    public List<List<Integer>> call() {
        // This produces a list of class indices of size numInts
        List<Integer> chosenClasses = getRandom()
                .ints(getNumInts(), 0, numBins)
                .boxed().toList();

        List<List<Integer>> outputList = new ArrayList<>();
        int numSets = (int) Math.ceil((double) getNumInts() / getSetSize());
        for (int i = 0; i < numSets - 1; i ++) {
            List<Integer> innerList = new ArrayList<>();
            for (int j = 0; j < getSetSize(); j ++) {
                int chosenClass = chosenClasses.get(i * getSetSize() + j);
                List<Integer> classList = classMap.get(chosenClass);
                int chosenInt = classList.get(
                        getRandom().nextInt(classList.size()));
                innerList.add(chosenInt);
            }
            outputList.add(innerList);
        }
        List<Integer> finalList = new ArrayList<>();
        for (int k = (numSets - 1) * getSetSize(); k < getNumInts(); k ++) {
            int chosenClass = chosenClasses.get(k);
            List<Integer> classList = classMap.get(chosenClass);
            int chosenInt = classList.get(
                    getRandom().nextInt(classList.size()));
            finalList.add(chosenInt);
        }
        outputList.add(finalList);
        return outputList;
    }

    /** Getter for {@code classMap}. Mainly for testing purposes.
     *
     * @return A copy of the {@code classMap}.
     */
    Map<Integer, List<Integer>> getClassMap() {
        Map<Integer, List<Integer>> returnMap = new HashMap<>();
        for (Integer k : classMap.keySet()) {
            returnMap.put(k, List.copyOf(classMap.get(k)));
        }
        return returnMap;
    }
}

File: pyNeuralNetwork/neural_network/util/weighted_partitioner.py
import math
import random
from typing import List
import pandas as pd

from .partitioner import Partitioner


class WeightedPartitioner(Partitioner):
    """Class to create `m` sets from a list of `n` integers weighted by which
    ground truth class each integer lies in.
    """

    def __init__(self, n: int, m: int, df: pd.DataFrame,
                 do_regression: bool = False, num_bins: int = 10):
        """Constructor method

        Parameters
        ----------
        n : int
            Number of integers
        m : int
            Number of sets for the partition
        df : pd.DataFrame
            The classes for the integers
        do_regression : bool
            Whether we are partitioning regressional or classificational data
        num_bins : int
            If regression is True, this represents the number of bins to
            split the data in to. Otherwise, this parameter is ignored
        """
        super().__init__(n, m)
        if not len(df) == n:
            raise ValueError(f"n must equal the length of the dataframe "
                             f"(n = {n}, len(df) = {len(df)})")
        if do_regression:
            self._num_bins = num_bins
            class_dict = {j: [] for j in range(self._num_bins)}
            min_y, max_y = min(df['y']), max(df['y'])
            class_width = (max_y - min_y) / num_bins
            for i in range(len(df)):
                y = df.loc[i, 'y']
                chosen_bin = int((y - min_y) / class_width)
                if chosen_bin == self._num_bins:
                    chosen_bin = self._num_bins - 1
                class_dict[chosen_bin].append(i)

            # Need to account for potentially empty class lists
            for j in range(self._num_bins):
                if not class_dict[j]:
                    class_dict.pop(j)
                    self._num_bins -= 1

            self._class_dict = {}
            for i, j in enumerate(class_dict.keys()):
                # Relabelling the classes
                self._class_dict[i] = class_dict[j]
        else:
            self._num_bins = len(set(df['y'].to_numpy()))
            self._class_dict = {j: [] for j in range(self._num_bins)}
            for i in range(len(df)):
                self._class_dict[int(df.loc[i, 'y'])].append(i)

    def __call__(self) -> List[List[int]]:
        """Uses weights for each class to create sets of size `m` containing
        integers (sampled with replacement).

        Returns
        -------
        List[List[int]]
            The list of sets
        """
        # Produces a list of `n` class indices
        chosen_classes = random.choices(population=range(self._num_bins),
                                        k=self._n)
        output_list = []
        num_sets = math.ceil(self._n / self._m)
        for i in range(num_sets - 1):
            inner_list = []
            for j in range(self._m):
                chosen_class = chosen_classes[i * self._m + j]
                inner_list.append(random.choice(
                    self._class_dict[chosen_class]))
            output_list.append(inner_list)
        inner_list = []
        for k in range((num_sets - 1) * self._m, self._n):
            chosen_class = chosen_classes[k]
            inner_list.append(random.choice(self._class_dict[chosen_class]))
        output_list.append(inner_list)
        return output_list
